### PR TITLE
drivers: modem: Adding SMS messaging support

### PR DESCRIPTION
**Pull request by johnlange2**
Adding function pointers for sending and receiving SMS messages
to the modem context. The receive SMS function may be blocking.

Signed-off-by: John Lange <John.Lange2@T-Mobile.com>

### CODE REVIEW

**Comment by johnlange2** [comment]
**Documentation for SMS Messaging Support in Zephyr**

Background: SMS (Short Message Service) is defined in RFC 5724

Below are some notes about the implementation:
-a new file, modem_sms.h, defines SMS sizes, structs, and functions
-two new function members were added to struct modem_context
-SMS support is modem-specific (must be implemented per modem)
-SMS receive is not guaranteed to return the oldest message first
-SMS receive is non-blocking if timeout is K_NO_WAIT
-SMS send/receive can be invoked using a modem context or a socket

Below are some examples:

Example 1: Send SMS message using modem context

```
#include <stddef.h>
#include <errno.h>
#include "modem_context.h"

int modem_send_sms(const int modem_index, const struct sms_out *sms)
{
	struct modem_context *ctx = modem_context_from_id(modem_index);

	if (ctx->send_sms == NULL) {
		return -ENOSYS;
	}

	return ctx->send_sms(ctx, sms);
}
```

Example 2: Receive SMS message using modem context

```
#include <stddef.h>
#include <errno.h>
#include "modem_context.h"

int modem_recv_sms(const int modem_index, struct sms_in *sms)
{
	struct modem_context *ctx = modem_context_from_id(modem_index);

	if (ctx->recv_sms == NULL) {
		return -ENOSYS;
	}

	return ctx->recv_sms(ctx, sms, K_NO_WAIT);
}
```

Example 3: Send SMS message using a socket

```
#include <fcntl.h>
#include "modem_sms.h"

int sms_send_on_socket(const int sock_fd, const struct sms_out *sms)
{
	// Call fcntl to send the SMS message
	return fcntl(sock_fd, SMS_SEND, &sms);
}

```
Example 4: Receive SMS message using a socket

```
#include <fcntl.h>
#include "modem_sms.h"

int sms_receive_on_socket(const int sock_fd, struct sms_in *sms)
{
	// Call fcntl to recv an SMS message
	return fcntl(sock_fd, SMS_RECV, sms, K_NO_WAIT);
}
```

**Review by rerickson1 — DISMISSED** [review]
Looks good!

File: drivers/modem/modem_context.h [hunk 1]
```diff
@@ -19,15 +20,16 @@
...
 extern "C" {
 #endif
 
 #define MODEM_PIN(name_, pin_, flags_) { \
-	.dev_name = name_, \
-	.pin = pin_, \
-	.init_flags = flags_ \
+		.dev_name = name_,	 \
+		.pin = pin_,		 \
+		.init_flags = flags_	 \
```
Comment (jukkar) [inline drivers/modem/modem_context.h:32]: The whitespace changes here and below are fine but should be placed in separate commit as they are not really related to SMS changes.

File: drivers/modem/modem_context.h [hunk 2]
```diff
@@ -67,12 +69,12 @@ struct modem_context {
...
-	int   data_operator;
-	int   data_lac;
-	int   data_cellid;
+	int data_operator;
+	int data_lac;
+	int data_cellid;
 #endif
 	int   *data_rssi;
-	bool  is_automatic_oper;
+	bool is_automatic_oper;
```
Comment (jukkar) [inline drivers/modem/modem_context.h:77]: ditto

File: drivers/modem/modem_context.h [hunk 3]
```diff
@@ -83,6 +85,10 @@ struct modem_context {
 	/* command handler config */
 	struct modem_cmd_handler cmd_handler;
 
+	/* SMS functions */
+	send_sms_func send_sms;
+	recv_sms_func recv_sms;
```
Comment (jukkar) [inline drivers/modem/modem_context.h:90]: The naming of these function pointers in other part of the system seems to be `xxxx_func_t` so I propose we follow the same naming convention.

File: drivers/modem/modem_shell.c [hunk 4]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+	struct sms_out sms;
+	int ret= -1;
+
+	if (argc != 4) {
+		shell_fprintf(shell, SHELL_ERROR,
+			      "usage: modem sms send <modem index> <phone number> <message>\n");
+	}
+	else {
+		int modem_index = atoi(argv[1]);
+		ms_ctx = ms_context_from_id(modem_index);
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:260]: Please put one empty line after variable definitions (to be consistent with other part of the system)

File: drivers/modem/modem_shell.c [hunk 5]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+		if (ms_ctx == NULL) {
+			shell_fprintf(shell, SHELL_ERROR,
+				      "Couldn't open context for modem index %d\n", modem_index);
+			return -1;
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:264]: At least in the network shell, the `-ENOEXEC` is returned in case of an error. Don't know if it makes much difference here.

File: drivers/modem/modem_shell.c [hunk 6]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+		}
+
+		if (ms_ctx->send_sms == NULL) {
+			shell_fprintf(shell, SHELL_ERROR,
+				      "No function defined to send SMS\n");
+			return -1;
+		}
+
+		snprintf(sms.phone, sizeof(sms.phone), "%s", argv[2]);
+		snprintf(sms.msg, sizeof(sms.msg), "%s", argv[3]);
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:274]: I propose `snprintk` instead, then we do not need to include stdio.h and I recall that the snprintk will also use less memory

File: drivers/modem/modem_shell.c [hunk 7]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+
+		ret = ms_ctx->send_sms(ms_ctx, &sms);
+
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:277]: Empty line here can be removed, our coding style puts variable setting and the corresponding `if` statement together.

File: drivers/modem/modem_shell.c [hunk 8]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+	struct ms_context *ms_ctx;
+	struct sms_in sms;
+	int ret = -1;
+
+	if (argc != 2) {
+		shell_fprintf(shell, SHELL_ERROR,
+			      "usage: modem sms recv <modem index>\n");
+	}
+	else {
+		int modem_index = atoi(argv[1]);
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:300]: empty line after variable declaration

File: drivers/modem/modem_shell.c [hunk 9]
```diff
@@ -242,11 +245,114 @@ static int cmd_modem_info(const struct shell *shell, size_t argc, char *argv[])
...
+		}
+
+		if (ms_ctx->recv_sms == NULL) {
+			shell_fprintf(shell, SHELL_ERROR,
+				      "No function defined to receive SMS\n");
+			return -1;
+		}
+
+		ret = ms_ctx->recv_sms(ms_ctx, &sms, K_NO_WAIT);
+
```
Comment (jukkar) [inline drivers/modem/modem_shell.c:315]: remove empty line

File: drivers/modem/modem_context.h [hunk 10]
```diff
@@ -7,6 +7,7 @@
```
Comment (jukkar) [inline drivers/modem/modem_context.h:7]: The shell changes should be placed in a separate commit in this PR

File: drivers/modem/modem_sms.h [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ * Modem SMS handling for modem driver.
+ */
+
+/*
+ * Copyright (c) 2021 T-Mobile USA, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
```
Comment (jukkar) [inline drivers/modem/modem_sms.h:13]: What is the purpose of this pragma, why do we need it?

**Review by jukkar — CHANGES_REQUESTED** [review]
Minor comments added. Could you create a sample app that shows how the API is used? I know there is the shell that shows this too but typically we have a sample app for that purpose as then it gets included in the documentation etc.

File: drivers/modem/modem_sms.h (hunk 11 above)
Comment (johnlange2) [reply drivers/modem/modem_sms.h:13]: From [https://en.wikipedia.org/wiki/Pragma_once](https://en.wikipedia.org/wiki/Pragma_once), "In the C and C++ programming languages, pragma once is a non-standard but widely supported preprocessor directive designed to cause the current source file to be included only once in a single compilation". Basically, it is used instead of the three lines: #ifndef and #define at the top, and #endif at the bottom.

Comment (jukkar) [reply drivers/modem/modem_sms.h:13]: Please, do not do that. Just put ifndef+define+endif like the other include files in the system.

File: drivers/modem/modem_context.h (hunk 2 above)
Comment (johnlange2) [reply drivers/modem/modem_context.h:77]: Done

File: drivers/modem/modem_shell.c (hunk 4 above)
Comment (johnlange2) [reply drivers/modem/modem_shell.c:260]: Done

File: drivers/modem/modem_shell.c (hunk 8 above)
Comment (johnlange2) [reply drivers/modem/modem_shell.c:299]: Done

File: drivers/modem/modem_context.h (hunk 10 above)
Comment (johnlange2) [reply drivers/modem/modem_context.h:7]: Done

File: drivers/modem/Kconfig [hunk 12]
```diff
@@ -139,6 +139,24 @@ config MODEM_CELL_INFO
...
+        default 160
+        help
+          This is the maximum length of inbound (received) SMS messages.
+          Inbound SMS messages longer than this will be truncated.
+          The absolute maximum value may be network- and/or modem-dependent,
+          but a smaller value may be preferred in order to conserve memory.
+
+config MODEM_SMS_OUT_MSG_MAX_LEN
+        int "Maximum length of outbound SMS messages"
+        default 160
```
Comment (jukkar) [inline drivers/modem/Kconfig:153]: I suppose the typical scenario is that both the inbound and outbound values are the same. So in order to avoid possible mismatch we could have here `default MODEM_SMS_IN_MSG_MAX_LEN`. Then user could override it if needed, what do you think?

File: drivers/modem/modem_sms.h [hunk 13]
```diff
@@ -16,27 +16,19 @@
...
-#endif
-
 /*
  * All fields in sms_out and sms_in are NULL terminated
  */
 
 struct sms_out {
         char phone[SMS_PHONE_MAX_LEN];
-        char msg  [CONFIG_SMS_OUT_MSG_MAX_LEN];
+        char msg  [CONFIG_MODEM_SMS_OUT_MSG_MAX_LEN + 2];
```
Comment (jukkar) [inline drivers/modem/modem_sms.h:25]: Why +2, wouldn't +1 be enough?

File: drivers/modem/Kconfig [hunk 14]
```diff
@@ -139,6 +139,24 @@ config MODEM_CELL_INFO
 	help
```
Comment (jukkar) [inline drivers/modem/Kconfig:139]: BTW, we do not do merge branches in zephyr history so please rebase the PR.

Comment (johnlange2) [reply drivers/modem/Kconfig:139]: Will do.

File: drivers/modem/modem_sms.h [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+struct sms_in {
+        char phone[SMS_PHONE_MAX_LEN];
+        char time [SMS_TIME_MAX_LEN];
+        char msg  [CONFIG_MODEM_SMS_IN_MSG_MAX_LEN + 2];
+};
+
+typedef int (*send_sms_func_t)(void *obj, const struct sms_out *sms);
+typedef int (*recv_sms_func_t)(void *obj,       struct sms_in  *sms, k_timeout_t timeout);
```
[the remark this answers is not in the document]
Comment (johnlange2) [reply drivers/modem/modem_sms.h:35]: Assuming no objections, I'll add the timeout to struct sms_in

**Review by ycsin — CHANGES_REQUESTED** [review]
I think not all modem will/can implement this function, i.e. wifi modems, so probably we need a Kconfig to prevent those modem from compiling the sms part (context, shell), or at least perform NULL check so that the shell wont invoke a null function

**Comment by ycsin** [comment]
Offtopic
@rlubos @jukkar @rerickson1 do you think if the RTC local time query can be done similarly like this PR?
